rfcs/20190802-model-garden-redesign.md

```diff
@@ -0,0 +1,207 @@
+APIs in different coding styles and some of them have convergence and/or
+performance regression. With TensorFlow 2.0 launch, there’s a great desire to
+provide tensorflow users a clear and central place to showcase reliable TF2.0
+models with the best practices to follow.
+
+We want to take this opportunity to substantially improve the state of the
+official model garden, and provide seamlessly end-to-end training and inference
+user experience on a wide range of accelerators and mobile device chips. We hope
+to encourage community to contribute innovations and improve TensorFlow
+efficiency and usability.
+
+## User Benefit
+
+We aim to provide the best modeling experience via this revamp effort:
+
+*   Usability and reliability
+    *   keep official models well-maintained and tested for both performance and
+        convergence.
+    *   provide accessible model distribution via [TensorFlow Hub](https://www.tensorflow.org/hub) and share state-of-the-art research accomplishments.
+    *   make training on both GPU and TPU an easy switch.
+    *   provide reusable components for research and production.
+*   End-to-end solutions
+    *   provide seamless end-to-end training and inference solutions, where inference covers serving on TPU, GPU, mobile and edge devices.
+    *   provide hyper parameter sets to tune models for various resource constraints.
+    *   provide solutions with hyper parameters to scale model training to TPU pods or multi-worker GPUs.
+    *   provide variants derived from standard models to tackle various practical tasks.
+
+## Design Proposal
+
+### Official model directory reorgnization
+
+We are going to reorganize the official model directory to provide:
+
+*   common libraries, mainly two types:
+    *   Common training util library in TF2.0, model configuration and
+        hyperparameter definition in a consistent style.
+    *   Model category related common library, e.g. primitives as basic building
+        block for NLP models. We will follow the fundamental design of Keras
+        layer/network/model to define and utilize model building blocks.
+*   popular state-of-the-art (SOTA) models for end users as a product.
+*   reference models for performance benchmark testing.
+    *   For models provided as SOTA models, we will share the network and
+        modeling code, but have separate *main* modules. The main
+        module for benchmark testing will have addtional flags and setups for
+        performance testing.
+
+The following table shows the detailed view of proposed model directory
+structure. The SOTA model list will be updated to cover more categories.
+
+| Directory     | Subdirectories       |   | Explainations                   |
+:-------------- |:---------------------|:--|:------------------------------ |
+| modeling      |                      |   | Common modeling libraries       |
+|               | layers               |   | Common modules/layers, not built-in tensorflow layers yet.|
```

Comment:
Splintering the TF ecosystem with multiple repos that contain new layers seems counter-productive. Addons is already going to manage the burden of graduating layers into core so why have 2 repos for the same thing. 

cc @karmel @facaiy for visibility

Reply:
With this, we are going in the direction of TF1.x again.

Reply:
@rachellj218 , I believe we discussed having a process to ensure any broadly useful layers/etc make it to Addons and/or tf.text. We should update the doc to reflect that, as it seems to be a common concern-- maybe a section describing what you would like the evaluation/graduation process to be?

Reply:
Thanks for raising this! As karmel@ mentioned, we do have the plan to graduate the common layers to tensorflow/addons. I clarified in the RFC. Thanks!

I removed optimizers/ subdirectory. I agree it should be good to add to Addons directly.

Reply:
Thanks for your clarification! In the latest RFC:
> Temporary folder to hold common modules/layers, not built-in tensorflow layers yet during refactoring. Broadly used layers will be graduated to tensorflow/addons and/or tf.text.

It seems to me that *graduation* will be another SLIM. The *graduation* will create different versions of one network which are based on different API, say a network uses 3 layers A, B and C firstly introduced in model garden, these 3 layers move to TF Addons one by one, thus we have four versions of network code. That is pretty confusing. And, as the interfaces of these layers are probably going to be different, it won't be as easy as re-write one line. Eventually, the effort to maintain these code will be huge.

What about implement the layers **directly** in TF Addons, which I think is much more flexible than the TF repo?

Thanks, hoping that my concern won't be too annoying...

Reply:
I agree with this concern. IMO it's less of a headache if they are moved from the temporary folder to Addons prior to a model-garden pip release. We can promise a quick review and patch release for official model additions. To ensure this we could have a model-garden team member on addons team.

Reply:
I think we need a consolidation stage for modeling components. This is not to develop infra but write models. We will guarantee layers moved to tf-addon must be removed in model garden. It is also up to how do you define "layers" as some of them are not common components as the ones in tf-addon now.

Reply:
> To ensure this we could have a model-garden team member on addons team.

+1, good idea. We can work in close coordination. 

Maybe model garden can put its private layers in each model's module. And when you find that some layers need be shared by two models or more, I believe it's a good time to move those layers to tf.addons, rather than modeling.layers. What do you think?